### PR TITLE
limiting number of blocks / chunks per height + adversary tests

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/1917.
Contains tests which should be fixed by doomslug and https://github.com/nearprotocol/nearcore/issues/1936.

### CODE REVIEW

File: pytest/tests/adversarial/malicious_chain.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+from cluster import start_cluster
+from adversary import corrupt_node
+from utils import LogTracker
+
+TIMEOUT = 300
+BLOCKS = 25
+MALICIOUS_BLOCKS = 150
+LARGE_WEIGHT = '1267650600228229401496703205376'
```
Comment (SkidanovAlex) [inline pytest/tests/adversarial/malicious_chain.py]: remove?

File: pytest/tests/adversarial/malicious_chain.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+nodes = start_cluster(2, 1, 2, None, [["epoch_length", MALICIOUS_BLOCKS + 5], ["block_producer_kickout_threshold", 80]], {})
+
+started = time.time()
+
+nodes[1].kill()
+nodes[2].kill()
```
Comment (SkidanovAlex) [inline pytest/tests/adversarial/malicious_chain.py]: remove?

File: chain/chunks/src/lib.rs [hunk 3]
```diff
@@ -590,20 +590,41 @@ impl ShardsManager {
...
+            if chain_store
+                .get_any_chunk_hash_by_height_shard(
+                    header.inner.height_created,
+                    header.inner.shard_id,
+                )
+                .is_ok()
+            {
+                info!(target: "client", "Reject unrequested chunk {:?}, height {}, shard_id {}", chunk_hash, header.inner.height_created, header.inner.shard_id);
+                return Err(Error::ChainError(
+                    ErrorKind::Other("Unrequested partial_encoded_chunk".to_string()).into(),
```
Comment (SkidanovAlex) [inline chain/chunks/src/lib.rs]: Create a new error kind?
Better "DuplicateChunkHeight" then "Unrequested"

File: chain/client/src/client_actor.rs [hunk 4]
```diff
@@ -165,20 +165,31 @@ impl Handler<NetworkClientMessages> for ClientActor {
...
                             }
-                            return NetworkClientResponses::NoResponse;
-                        } else if &block.hash() == sync_hash {
-                            self.client.chain.save_orphan(&block);
-                            return NetworkClientResponses::NoResponse;
                         }
                     }
+                    self.receive_block(block, peer_id, was_requested)
+                } else {
+                    info!(target: "client", "Reject unrequested block {}, height {}", block.header.hash, block.header.inner_lite.height);
```
Comment (SkidanovAlex) [inline chain/client/src/client_actor.rs]: Reject -> Rejecting

File: pytest/tests/adversarial/malicious_blocks.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+print("Got to %s blocks, getting to fun stuff" % BLOCKS)
+
+nodes[0].kill() # to disallow syncing
+nodes[1].kill()
+nodes[1].reset_data()
+#corrupt_node(nodes[0]) # for calling adv_get_saved_blocks
+
+#corrupt_node(nodes[1])
+#tracker = LogTracker(nodes[1])
```
Comment (SkidanovAlex) [inline pytest/tests/adversarial/malicious_blocks.py]: uncomment?

File: pytest/tests/adversarial/malicious_chain.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+TIMEOUT = 300
+BLOCKS = 25
+MALICIOUS_BLOCKS = 50
+
+nodes = start_cluster(2, 1, 2, None, [["epoch_length", 1000], ["block_producer_kickout_threshold", 80]], {})
+
+started = time.time()
+
+#corrupt_node(nodes[1]) # works if .adv_on_top, not tested otherwise, maybe should be killed and started again
```
Comment (SkidanovAlex) [inline pytest/tests/adversarial/malicious_chain.py]: if you corrupt, you need to kill before corruption and restart after
also, uncomment?

File: chain/chunks/src/lib.rs (hunk 3 above)
Comment (Kouprin) [reply chain/chunks/src/lib.rs]: ++

File: chain/client/src/client_actor.rs (hunk 4 above)
Comment (Kouprin) [reply chain/client/src/client_actor.rs]: ++

File: pytest/tests/adversarial/malicious_blocks.py (hunk 5 above)
Comment (Kouprin) [reply pytest/tests/adversarial/malicious_blocks.py]: uncommenting `corrupt_node` execution

Comment (Kouprin) [reply pytest/tests/adversarial/malicious_blocks.py]: removed tracker

File: pytest/tests/adversarial/malicious_chain.py (hunk 6 above)
Comment (Kouprin) [reply pytest/tests/adversarial/malicious_chain.py]: uncommenting `corrupt_node` execution

**Pull request merged**